pkg/ruler/manager.go

```diff
@@ -105,8 +106,20 @@ func (r *DefaultMultiTenantManager) SyncRuleGroups(ctx context.Context, ruleGrou
 		RemoveFederatedRuleGroups(ruleGroups)
 	}
 
-	for userID, ruleGroup := range ruleGroups {
-		r.syncRulesToManager(ctx, userID, ruleGroup)
+	// Sync the rules to disk and then update the user's Prometheus Rules Manager.
+	// Since users are different, we can sync rules in parallel.
+	users := make([]string, 0, len(ruleGroups))
+	for userID := range ruleGroups {
+		users = append(users, userID)
+	}
+	err := concurrency.ForEachJob(ctx, len(users), 10, func(ctx context.Context, idx int) error {
```

Comment:
As you can see all ruler integration tests fail. The reason is that the context is preserved by the manager created in `r.syncRulesToManager()` but `concurrency.ForEachJob()` cancels the child context passed here once the iteration is over, so this cancels the context retained by the manager.

What we want here is to use the parent context and not the child one from `concurrency.ForEachJob()`:

```suggestion
	err := concurrency.ForEachJob(ctx, len(users), 10, func(_ context.Context, idx int) error {
```

Please also add a comment to clearly explain why we do ☝️